src/core/qgssqlexpressioncompiler.cpp

```diff
@@ -93,6 +93,42 @@ QgsSqlExpressionCompiler::Result QgsSqlExpressionCompiler::compileNode( const Qg
   if ( staticRes != Fail )
     return staticRes;
 
+  // This is just to identify the most simple cases where nodes are numeric
+  std::function<bool( const QgsExpressionNode * )> nodeIsNumeric;
+  nodeIsNumeric = [this, &nodeIsNumeric]( const QgsExpressionNode * node )
+  {
+    const QgsExpressionNode::NodeType nodeType { node->nodeType() };
+
+    switch ( nodeType )
+    {
+      case QgsExpressionNode::ntColumnRef:
+      {
+        const QgsExpressionNodeColumnRef *col = static_cast<const QgsExpressionNodeColumnRef *>( node );
+        const int idx = mFields.indexFromName( col->name() );
+        return idx >= 0 && mFields[idx].isNumeric();
+      }
+      case QgsExpressionNode::ntLiteral:
+      {
+        const QgsExpressionNodeLiteral *lit = static_cast<const QgsExpressionNodeLiteral *>( node );
+        return lit->value().type() == QVariant::Double || lit->value().type() == QVariant::Int;
```

Comment:
How about moving the variant type comparison logic from QgsField::isNumeric to something like QgsVariantUtils::isNumericType and using here? Otherwise you'll also need to add checks manually here for LongLong, ULongLong & Uint

Reply:
I have opted for QgsVariantUtils::isNumericType but I am not sure about how to handle bool an Char, can you please give me an advice? See https://github.com/qgis/QGIS/pull/58916/commits/61935ad591d7cf34ba1b7791536f1a434d8e30b2#diff-ce9de7392278448888de27a62f580122579719433636c7e0aed544a0c7004549R485 and the test.

Reply:
I'd lean toward false there (my gut feeling is that the more common use case would be storing a single text character rather than a number), but don't have a strong opinion.

Reply:
I'll approve as is